### ISSUE TITLE
Problem with not touched conf.d after update to 2.11

### ISSUE THREAD

**Issue by nazco** [Icinga/icinga2#7572]
Well this is a rather strange bug in my opinion and we see it all over our cluster but only for agents upgraded to 2.11. We use a distributed setup with on master three satellites and a lot of clients. They are configured via the director. The distributed setup is build according to the docs. The zones where created via the director too (as suggested by the director itself), despite the master zone.

----
Despite that we seem to be affected (in some way I guess?) by #7530 or #7514  as the director 
mentioned that... but I have no Idea what the problem should be, a better documentation of the needed structure would be great or how I can fix it.
![Screenshot2019-10-07_17-00-34](https://user-images.githubusercontent.com/1292551/66639552-4d7dbe00-ec17-11e9-9a3e-dae54fd06124.png)
But this is another problem because the upgrade docs say "if you are using the director you are save [of this bug]"

----
Back to the problem of this issue (I mention the stuff before because I think there could be a connection). After upgrading a node we get a warning from our icinga agent check:

> Icinga 2 has been running for 8 days, 6 hours, 11 minutes and 47 seconds. Version: 2.11.0-1; Last zone sync stage validation failed at 2019-10-10 18:10:46 +0200

The startup log describes a command_endpoint error. So the new config isn't loaded.

```
[2019-10-10 18:10:09 +0200] information/cli: Icinga application loader (version: 2.11.0-1)
[2019-10-10 18:10:09 +0200] information/cli: Loading configuration file(s).
[2019-10-10 18:10:09 +0200] information/ConfigItem: Committing config item(s).
[2019-10-10 18:10:09 +0200] information/ApiListener: My API identity: dietrich.example.com
[2019-10-10 18:10:09 +0200] critical/config: Error: Validation failed for object 'dietrich.example.com!Icinga Agent' of type 'Service'; Attribute 'command_endpoint': Command endpoint must not be set.
Location: in /var/lib/icinga2/api/zones-stage//director-global/director/service_apply.conf: 33:1-33:28
/var/lib/icinga2/api/zones-stage//director-global/director/service_apply.conf(31): }
/var/lib/icinga2/api/zones-stage//director-global/director/service_apply.conf(32): 
/var/lib/icinga2/api/zones-stage//director-global/director/service_apply.conf(33): apply Service "Icinga Agent" {
                                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/var/lib/icinga2/api/zones-stage//director-global/director/service_apply.conf(34):     import "Icinga"
/var/lib/icinga2/api/zones-stage//director-global/director/service_apply.conf(35): 

[2019-10-10 18:10:09 +0200] critical/config: 1 error
[2019-10-10 18:10:09 +0200] critical/cli: Config validation failed. Re-run with 'icinga2 daemon -C' after fixing the config.
```
After debugging that for some hours we tried to reconnect on of the agents with the node wizard. After that the problem was gone. Strange. After a check what has change we observed that the only thing that changed was the exclusion of the conf.d directory (which we never ever touched, it's the original coming from your repos packages). So we disabled only the inclusion of the conf.d directory on another agent host. And the error was gone too. No agent warning no problems in the startup log. And now we have no idea why the conf.d directory we never touched triggered that errors. Maybe you have an idea?

Oh and clients with the latest 2.10 release are not affected by that problem.

## Your Environment
Version: 2.11
Operating System and version: Debian 10, Debian 9 and CentOS 7

**Comment by dnsmichi** [comment]
I'm doing my best to update the docs for that Zone-in-Zone thingy, if you have an idea how to write it better, please suggest or send in a PR. Lately I've updated the linked issue from the Director, but right now this is spread all over the place. Not happy with it yet, it has to wait though until my other tasks are finished.

In terms of the problem with conf.d - it seems that your upgrade did override `icinga2.conf` and re-enable the inclusion of that directory. The CLI wizards since 2.10 or so disable this entry by default, like 

```
// Disabled by the node setup CLI command on 2019-07-31 14:26:15 +0200
// include_recursive "conf.d"
```

Whenever you re-run the node wizard CLI tool, this inclusion will vanish with the commenting out as described. 

Debian asks whenever it wants to override config files, RHEL keeps a .rpmnew file or so but does not touch the existing one. There might be insights in apt/yum history log on why these configuration file has been overwritten.

On the long run, to avoid errors like this, I'd suggest to use the power of Puppet and Ansible and always ensure that the configuration stays the same, and also to roll upgrade changes automatically.

That also allows you to adopt the zones.conf file in a way to escape the zone inception problem now, since it is managed within Puppet/Ansible and only managed by the Director as externally defined object.

Cheers,
Michael

**Comment by nazco** [comment]
> In terms of the problem with conf.d - it seems that your upgrade did override icinga2.conf and re-enable the inclusion of that directory.

We originally came from a version where the conf.d directory was not excluded. And our automation doesn't touch the icinga2.conf. Also there is no .rpnnew file (clients/agents are all CentOS). So the problem has to be another one.

>On the long run, to avoid errors like this, I'd suggest to use the power of Puppet and Ansible and always ensure that the configuration stays the same, and also to roll upgrade changes automatically.

Well we use ansible, and the problem would have also happened if we ensured the original version of icinga2.conf because your upgrade notes never said that you have to exclude it to work properly. And yes we also roll upgrades automatically. Thats the reason why our whole cluster is affected. After the installation and registration process, which happens with ansible, we don't touch the agent locally. So I don't get what problem you see :)

> That also allows you to adopt the zones.conf file in a way to escape the zone inception problem now, since it is managed within Puppet/Ansible and only managed by the Director as externally defined object.

 I'm still not sure what you want to tell me here? The zones are only managed via the director. Or do you mean to remove the master zone from the zones.conf and readd it in the director (which I think will fail as of some constrains)

**Comment by dnsmichi** [comment]
Wasn't aware of that how much the Director pushes you into this direction, sorry. I'm slowly adopting the Director into my own workflows, Icinga 2 was designed long before the Director actually existed and I truly am a beginner with the Director myself. If that irritates you, no worries. We're getting better here every day, also from this issue.

@lippserd @bobapple we need to review the feature set here imho. We for Icinga 2 always encouraged users to define the zones.d inclusion in zones.conf to prevent the chicken-egg problem. It is really unfortunate that the Director proposes this in another way, leading to these irritations. No-one's to blame here, I'm eager to find the best solution and documentation how to build Icinga environments.

> because your upgrade notes never said that you have to exclude it to work properly.

Hmmm, I don't remember when we suggested users to not use the local configuration, but I am pretty sure this was mentioned somewhere in the Changelog, like https://icinga.com/docs/icinga2/latest/doc/16-upgrading-icinga-2/#setup-wizard-changes 
Still, I get your point to make this more explicit for those using automation tools.

That being said, I can only say sorry for the change with the command_endpoint and conf.d. Our tests have shown that this works for us, but we did not see nor expect the impact of this change. That is why this change wasn't highlighted as explicit required user step in the upgrading docs. I wish I could turn back time and do it better.

I've updated the linked issue with a fresh zones.conf btw.

Cheers,
Michael

**Comment by Thomas-Gelf** [comment]
Hi @nazco,

we're sorry that you have been bitten by this change. Let me try to shine some light on what happened in your setup. You mentioned that you never touched `icinga2.conf` on your agents. That's where your problem originates, as **your configuration was flawed from the beginning**. Strange things where going on, but the problem was neither visible nor obivios. What changed with 2.11 is that right now your Agents refuse to start.

> Including `conf.d` is and always used to be a problem, as it contains sample configuration that could easily conflict with distributed configuration as shipped by the Director or other tools. That's why all Agent deployment instructions and utilities offered by the Director create and propose configurations not including `conf.d` at all. Recently, also `icinga2 node wizard` suggests to not include that directory.

So, what was wrong with your setup? As most other people, you centrally deployed a configuration with all Host and Service objects sitting in your Agent's parent zones. Per default, Director ships no configuration but global zones to your Agents. All checks are executed via the `command_endpoint` mechanism, scheduling takes place on your Master and Satellite nodes.

However, when your Agents locally include `conf.d`, they suddenly get their very own configuration. They define a Host object corresponding to their own name, and some services will be applied. Also, `Apply Rules` in the global zone shipped by the Director might match. They run kind of a shadow configuration, visible nowhere else.

Chances are good that your setup kept happily running this way the whole time. Your agents accomplished some useless/duplicate work with nobody noticing that they did so. And everything looked good.

Now, what changed with 2.11? Icinga refuses to start on your Agents and tries to point you to the error. **This is annoying, but helps to clean up your setup**. All you have to do is to take control of `icinga2.conf` in your Ansible configuration. The default one is good for demo setups, but harmful in a distributed environment. You can use the Agent configurations proposed by the Director as a starting point.

Now what remains is the yellow notice Director shows in your setup before deploying the configuration. This refers to Zone and Endpoint objects you explicitly created via Director, those created automagically by the Director for your Agents are safe. Director always advised against doing so *unless you know what you're doing*, but this doesn't mean that doing so is technically wrong.

That this no longer works is a bug in 2.11, it will be fixed. Please follow the linked issue for related details. In the meantime you can either move all your Zone/Endpoint-Definitions into your local `zones.conf` files (that's tricky, please read [this](https://github.com/Icinga/icinga2/issues/7542#issuecomment-535976245)) or stay on 2.10.x unless this has been fixed.

Hope this helps,
Thomas

NB: One last point, just to avoid confusion:

> It is really unfortunate that the Director proposes this in another way, leading to these irritations.

@dnsmichi: on what evidence do you base this statement?

**Comment by dnsmichi** [comment]
> > It is really unfortunate that the Director proposes this in another way, leading to these irritations.
> 
> @dnsmichi: on what evidence do you base this statement?

This statement is really unfortunate, shown to the users encouraging them to use something which is not supported by Icinga 2.

> When you feel the desire to manually create Zone or Endpoint objects please rethink this twice. Doing so is mostly the wrong way, might lead to a dead end, requiring quite some effort to clean up the whole mess afterwards.

Following the discussion above with 

> I'm still not sure what you want to tell me here? The director even heavily suggests me to create the zones in the director and not manually...

leads me to the assumption that users think that they **should** create Endpoint/Zone objects **inside the Director**, not outside how the Icinga 2 documentation suggests.

Imho the help text in https://github.com/Icinga/icingaweb2-module-director/blob/5404e32f6dbd7dbab373a321d09d69cae6072891/library/Director/Dashboard/InfrastructureDashboard.php#L28 needs to fixed making it more clear that certain configuration modes are not supported, e.g.

```
When managing Zone/Endpoint objects here, keep in mind that this currently is only supported for Host agents. Master/satellite zones should be managed outside but can be viewed here.
```

Cheers,
Michael

**Comment by Thomas-Gelf** [comment]
> This statement is really unfortunate, shown to the users encouraging them to use something which is not supported by Icinga 2.

I'm unable to decipher what you meant to say here, sorry.

> Following the discussion above with [..] leads me to the assumption that users think [..]

That assumption is plain wrong.

> Imho the help text [..] needs to fixed [..]

The help text reads:

> *When you feel the desire to manually create Zone or Endpoint objects please rethink this twice. Doing so is mostly the wrong way, might lead to a dead end, requiring quite some effort to clean up the whole mess afterwards.*

More obvious and direct than that? No, sorry. IMHO **the bug itself should be fixed**. That would be fantastic - and safe a lot of  time we currently spend on teaching people what they're doing wrong.

What @nazco has been running locally (by including `conf.d`) has always been wrong. What he did with his Satellite Zones is hard to grasp for the beginner, but still **absolutely valid** and the only sane option in a distributed environment.

**Comment by dnsmichi** [comment]
> More obvious and direct than that? No, sorry. IMHO the bug itself should be fixed. That would be fantastic - and safe a lot of time we currently spend on teaching people what they're doing wrong.

I'm not talking about the missing Zone-in-Zone feature in the core here, that's going to be a task in the coming weeks for a core developer. 

Instead, I am interested to get an idea on **why** users tend to use this without our knowledge. The help text mentioned above leads to irritations, I was just quoting @nazco here. To avoid confusion within our products, I'd suggest to modify the text. @lippserd your opinion on this?

**Comment by Thomas-Gelf** [comment]
> Instead, I am interested to get an idea on **why** users tend to use this without our knowledge

Probably also because it always worked unless we broke it with v2.11? Eventually because it is the most obvious way of handling things? Because you do not want to manage your nested zones in multiple configuration tools in parallel?

But who knows. We could start a poll, try to break this complexity down into a simple question and wait for meaningful answers. Or, but that's just my naive point of view: we could just fix the bug.

**Comment by dnsmichi** [comment]
The tone in this issue is way off, and imho not the way we should communicate with each other on a respectful level. 

That being said, I'll close this.

The documentation changes will follow in a separate PR.
